ERCS/erc-5559.md

```diff
@@ -1,471 +1,449 @@
 
-In step 1, a standard blockchain call operation is made to the contract. The contract either performs the operation as intended or reverts with an error that specifies the type of handler that the mutation is being deferred to and the corresponding parameters required to perform the subsequent mutation. There are two types of errors that the contract can revert with, but more may be defined in other EIPs:
+```solidity
+// Function in L2 contract
+function setValue(
+    bytes32 node,
+    bytes32 key,
+    bytes32 value
+) external {
+    // Some code storing data mapped by node & msg.sender
+    ...
+}
+```
 
-- `StorageHandledByL2(chainId, contractAddress)`
-- `StorageHandledByOffChainDatabase(sender, url, data)`
+### Database Handler: `StorageHandledByDatabase()`
+A minimal database handler is similar to an L2 in the sense that:
 
-In step 2, the client builds and performs a new request based off of the type of error received in (1). These handshakes are outlined in the sections below:
+  a) Similar to `chainId`, it requires the `gatewayUrl` that is tasked with handling off-chain write operations, and
 
-- [StorageHandledByL2](#data-stored-in-an-l2)
-- [StorageHandledByOffChainDatabase](#data-stored-in-an-off-chain-database) 
+  b) Similar to `eth_call`, it requires `eth_sign` output to secure the data, and the client must prompt the users for these signatures.
 
-In some cases, the mutation may be deferred multiple times
-- [Storage Deferred Twice L1 > L2 > Off-Chain](#data-stored-in-an-l2--an-off-chain-database) 
+This specification does not require any other data to be stored on L1 other than the bespoke `gatewayUrl`; the storage handler therefore should only return the `gatewayUrl` in revert.
 
-### Data Stored in an L1
-```
-┌──────┐                ┌───────────┐ 
-│Client│                │L1 Contract│ 
-└──┬───┘                └─────┬─────┘ 
-   │                          │       
-   │ somefunc(...)            │       
-   ├─────────────────────────►│       
-   │                          │       
-   │ response                 │       
-   │◄─────────────────────────┤       
-   │                          │       
+```solidity
+error StorageHandledByDatabase(
+    string gatewayUrl
+);
+
+// Generic function in a contract
+function setValue(
+    bytes32 node,
+    bytes32 key,
+    bytes32 value
+) external {
+    (
+        string gatewayUrl // Gateway URL; may be globally constant
+    ) = getMetadata(node);
+    // gatewayUrl = "https://api.namesys.xyz"
+    // Defer write call to database handler
+    revert StorageHandledByDatabase( 
+        gatewayUrl
+    );
+};
 ```
 
-In the case in which no reversion occurs, data is stored in the L1 contract when the transaction is executed.
+Following the revert, the client must take these steps:
 
-### Data Stored in an L2
+1. Request the user for a secret signature `sigKeygen` to generate a deterministic `dataSigner` keypair,
```

Comment:
> The only benefit I am able to comprehend is that your solution can provide optional password

I don't know what you mean by this quite honestly. Can you please rephrase? The benefits have nothing to do with `password` which is not a feature.

True benefits that you missed out start from line 200:

```
Since the derived signer is wallet-specific, it can 

- sign batch data for multiple keys for a given node, and 
- sign batches of data for multiple nodes owned by a wallet

simultaneously in the background without ever prompting the user.
```